### PR TITLE
Implemented Sawmill

### PR DESCRIPTION
**Pull request by LeviHarrison**
#57 this PR I've implemented Sawmill in almost all of the Functions, excluding `tokengen` and `updatealgolia` because those are weird. I've also added a spot in `config.yaml` for the credentials. I plan to add the Emoji improvements to Sawmill in another PR, just wanted to get this out.

### CODE REVIEW

**Comment by LeviHarrison** [comment]
Sorry about all the cryptic commit names, I accidentally left in some things in that shouldn't be in production.

File: cmd/appleauth/appleauth.go [hunk 1]
```diff
@@ -19,6 +20,7 @@ type appleDevTokenResp struct {
 }
 
 var firestoreClient *firestore.Client
+var logger sawmill.Logger
```
Comment (adilanchian) [inline cmd/appleauth/appleauth.go:23]: 🔥  sawmill.Logger 🔥 

LETS GO.

File: cmd/appleauth/appleauth.go [hunk 2]
```diff
@@ -35,7 +37,7 @@ func AuthorizeWithApple(writer http.ResponseWriter, request *http.Request) {
 	initWithEnvErr := initWithEnv()
 	if initWithEnvErr != nil {
 		http.Error(writer, initWithEnvErr.Error(), http.StatusInternalServerError)
-		log.Printf("[AuthorizeWithApple] [initWithEnv]: %v", initWithEnvErr)
+		logger.LogErr(initWithEnvErr, "initWithEnv", nil)
```
Comment (adilanchian) [inline cmd/appleauth/appleauth.go]: So I think in the `LogErr` method in `sawmill`, we should change the param order to be:

`operation, error, request`

Thoughts on this? I feel like there is a lot more readability in this type of layout.

Comment (adilanchian) [reply cmd/appleauth/appleauth.go]: [nit]

`initWithEnv` -> `InitWithEnv`

File: cmd/appleauth/appleauth.go [hunk 3]
```diff
@@ -44,14 +46,14 @@ func AuthorizeWithApple(writer http.ResponseWriter, request *http.Request) {
 	appleDevTokenReq, appleDevTokenReqErr := http.NewRequest("GET", createAppleDevURI, nil)
 	if appleDevTokenReqErr != nil {
 		http.Error(writer, appleDevTokenReqErr.Error(), http.StatusInternalServerError)
-		log.Println(appleDevTokenReqErr.Error())
+		logger.LogErr(appleDevTokenReqErr, "appleDevToken Request", appleDevTokenReq)
```
Comment (adilanchian) [inline cmd/appleauth/appleauth.go]: [nit]

`appleDevToken ` -> `AppleDevToken `

File: cmd/appleauth/appleauth.go [hunk 4]
```diff
@@ -44,14 +46,14 @@ func AuthorizeWithApple(writer http.ResponseWriter, request *http.Request) {
...
 		return
 	}
 
 	appleDevTokenResp, appleDevTokenRespErr := httpClient.Do(appleDevTokenReq)
 	if appleDevTokenRespErr != nil {
 		http.Error(writer, appleDevTokenRespErr.Error(), http.StatusInternalServerError)
-		log.Println(appleDevTokenRespErr.Error())
+		logger.LogErr(appleDevTokenRespErr, "appleDevToken Response", nil)
```
Comment (adilanchian) [inline cmd/appleauth/appleauth.go]: [nit]

`appleDevToken` -> `AppleDevToken`

File: cmd/appleauth/appleauth.go [hunk 5]
```diff
@@ -60,7 +62,7 @@ func AuthorizeWithApple(writer http.ResponseWriter, request *http.Request) {
 	appleDevTokenDecodeErr := json.NewDecoder(appleDevTokenResp.Body).Decode(&appleDevToken)
 	if appleDevTokenDecodeErr != nil {
 		http.Error(writer, appleDevTokenDecodeErr.Error(), http.StatusInternalServerError)
-		log.Printf("AppleAuth [appleDevToken Decoder]: %v", appleDevTokenDecodeErr)
+		logger.LogErr(appleDevTokenDecodeErr, "appleDevToken Decoder", nil)
```
Comment (adilanchian) [inline cmd/appleauth/appleauth.go]: [nit]

`appleDevToken` -> `AppleDevToken`

File: cmd/appleauth/appleauth.go [hunk 6]
```diff
@@ -71,7 +73,7 @@ func AuthorizeWithApple(writer http.ResponseWriter, request *http.Request) {
 	renderErr := render.HTML(writer, http.StatusOK, "auth", appleDevToken)
 	if renderErr != nil {
 		http.Error(writer, renderErr.Error(), http.StatusInternalServerError)
-		log.Printf("[AuthorizeWithApple] [render]: %v", renderErr)
+		logger.LogErr(renderErr, "render", nil)
```
Comment (adilanchian) [inline cmd/appleauth/appleauth.go]: [nit]

`render` -> `Render`

File: cmd/createappledevtoken/createappledevtoken.go [hunk 7]
```diff
@@ -31,15 +33,15 @@ func CreateAppleDevToken(writer http.ResponseWriter, request *http.Request) {
...
+		logger.LogErr(initWithEnvErr, "initWithEnv", nil)
 		return
 	}
 
 	// Check for developer token in firebase
 	devToken, devTokenErr := fetchToken()
 	if devTokenErr != nil {
 		http.Error(writer, devTokenErr.Error(), http.StatusInternalServerError)
-		log.Printf("[CreateAppleDevToken] [getAppleDevToken]: %v", devTokenErr)
+		logger.LogErr(devTokenErr, "getAppleDevToken", nil)
```
Comment (adilanchian) [inline cmd/createappledevtoken/createappledevtoken.go]: [nit]

`getAppleDevToken` -> `GetAppleDevToken`

File: cmd/createappledevtoken/createappledevtoken.go [hunk 8]
```diff
@@ -109,6 +111,13 @@ func initWithEnv() error {
...
 	}
 
+	// Initialize Sawmill
+	sawmillLogger, err := sawmill.InitClient(currentProject, os.Getenv("GCLOUD_CONFIG"), os.Getenv("ENVIRONMENT"), "CreateAppleDevToken")
+	if err != nil {
+		log.Printf("CreateAppleDevToken [Init Sawmill]: %v", err)
+	}
+
+	logger = sawmillLogger
 	firestoreClient = client
```
Comment (adilanchian) [inline cmd/createappledevtoken/createappledevtoken.go]: [nit]

In the other function you set `logger = sawmillLogger` under the `firestoreClient`. Can you change that here?

File: cmd/createprovideruser/createprovideruser.go [hunk 9]
```diff
@@ -23,14 +24,15 @@ type providerUser struct {
...
+var logger sawmill.Logger
 
 // CreateProviderUser will check to see if the newly created user needs to be added to the providers_users collection
 func CreateProviderUser(writer http.ResponseWriter, request *http.Request) {
 	// Initialize
 	initErr := initWithEnv()
 	if initErr != nil {
 		http.Error(writer, initErr.Error(), http.StatusInternalServerError)
-		log.Printf("CreateProviderUser [Init failure]: %v", initErr)
+		logger.LogErr(initErr, "initWithEnv", nil)
```
Comment (adilanchian) [inline cmd/createprovideruser/createprovideruser.go]: [nit]

`initWithEnv` -> `InitWithEnv`

File: cmd/createprovideruser/createprovideruser.go [hunk 10]
```diff
@@ -40,7 +42,7 @@ func CreateProviderUser(writer http.ResponseWriter, request *http.Request) {
 	reqDataErr := json.NewDecoder(request.Body).Decode(&reqData)
 	if reqDataErr != nil {
 		http.Error(writer, reqDataErr.Error(), http.StatusInternalServerError)
-		log.Printf("CreateProviderUser [reqData Decoder]: %v", reqDataErr)
+		logger.LogErr(reqDataErr, "reqData Decoder", request)
```
Comment (adilanchian) [inline cmd/createprovideruser/createprovideruser.go]: [nit]

`reqData ` -> `ReqData `

File: cmd/createprovideruser/createprovideruser.go [hunk 11]
```diff
@@ -57,7 +59,7 @@ func CreateProviderUser(writer http.ResponseWriter, request *http.Request) {
 	_, writeErr := firebaseProviderDocRef.Set(context.Background(), providerUserData)
 	if writeErr != nil {
 		http.Error(writer, writeErr.Error(), http.StatusInternalServerError)
-		log.Printf("CreateProviderUser [fireStore Set]: %v", writeErr)
+		logger.LogErr(writeErr, "fireStore Set", request)
```
Comment (adilanchian) [inline cmd/createprovideruser/createprovideruser.go]: [nit]

`fireStore -> FireStore`

File: cmd/createsocialplaylist/createsocialplaylist.go [hunk 12]
```diff
@@ -64,7 +66,7 @@ func CreateSocialPlaylist(writer http.ResponseWriter, request *http.Request) {
 	err := initWithEnv()
 	if err != nil {
 		http.Error(writer, err.Error(), http.StatusInternalServerError)
-		log.Printf("CreateSocialPlaylist [initWithEnv]: %v", err.Error())
+		logger.LogErr(err, "initWithEnv", nil)
```
Comment (adilanchian) [inline cmd/createsocialplaylist/createsocialplaylist.go]: [nit]

`initWithEnv` -> `InitWithEnv`

File: cmd/createsocialplaylist/createsocialplaylist.go [hunk 13]
```diff
@@ -74,7 +76,7 @@ func CreateSocialPlaylist(writer http.ResponseWriter, request *http.Request) {
 	jsonDecodeErr := json.NewDecoder(request.Body).Decode(&socialPlaylistReq)
 	if jsonDecodeErr != nil {
 		http.Error(writer, jsonDecodeErr.Error(), http.StatusInternalServerError)
-		log.Printf("CreateSocialPlaylist [socialPlaylistReq Decoder]: %v", jsonDecodeErr)
+		logger.LogErr(jsonDecodeErr, "socialPlaylistReq Decoder", request)
```
Comment (adilanchian) [inline cmd/createsocialplaylist/createsocialplaylist.go]: [nit]

`socialPlaylistReq ` -> `SocialPlaylistReq `

File: cmd/createsocialplaylist/createsocialplaylist.go [hunk 14]
```diff
@@ -92,7 +94,7 @@ func CreateSocialPlaylist(writer http.ResponseWriter, request *http.Request) {
 		socialRefreshTokens, socialRefreshTokenErr = refreshToken(socialPlaylistReq.SocialPlatform)
 		if socialRefreshTokenErr != nil {
 			http.Error(writer, socialRefreshTokenErr.Error(), http.StatusBadRequest)
-			log.Printf("CreateSocialPlaylist [refreshToken]: %v", socialRefreshTokenErr)
+			logger.LogErr(socialRefreshTokenErr, "refreshToken", request)
```
Comment (adilanchian) [inline cmd/createsocialplaylist/createsocialplaylist.go]: [nit]

`refreshToken` -> `RefreshToken`

File: cmd/createsocialplaylist/createsocialplaylist.go [hunk 15]
```diff
@@ -127,7 +129,7 @@ func CreateSocialPlaylist(writer http.ResponseWriter, request *http.Request) {
 	createReqErr := createPlaylist(platformEndpoint, socialPlaylistReq.SocialPlatform, socialPlaylistReq.PlaylistName)
 	if createReqErr != nil {
 		http.Error(writer, createReqErr.Error(), http.StatusBadRequest)
-		log.Printf("CreateSocialPlaylist [createPlaylist]: %v", createReqErr.Error())
+		logger.LogErr(createReqErr, "createPlaylist", request)
```
Comment (adilanchian) [inline cmd/createsocialplaylist/createsocialplaylist.go]: [nit]

`createPlaylist ` -> `CreatePlaylist `

File: cmd/createuser/createuser.go [hunk 16]
```diff
@@ -25,7 +27,7 @@ func CreateUser(writer http.ResponseWriter, request *http.Request) {
 	initWithEnvErr := initWithEnv()
 	if initWithEnvErr != nil {
 		http.Error(writer, initWithEnvErr.Error(), http.StatusInternalServerError)
-		log.Printf("CreateUser [initWithEnv]: %v", initWithEnvErr)
+		logger.LogErr(initWithEnvErr, "initWithEnv", nil)
```
Comment (adilanchian) [inline cmd/createuser/createuser.go]: [nit]

`initWithEnv ` -> `InitWithEnv `

File: cmd/createuser/createuser.go [hunk 17]
```diff
@@ -34,15 +36,15 @@ func CreateUser(writer http.ResponseWriter, request *http.Request) {
 	jsonDecodeErr := json.NewDecoder(request.Body).Decode(&createUserReq)
 	if jsonDecodeErr != nil {
 		http.Error(writer, jsonDecodeErr.Error(), http.StatusInternalServerError)
-		log.Printf("CreateUser [social.CreateUserReq Decoder]: %v", jsonDecodeErr)
+		logger.LogErr(jsonDecodeErr, "social.CreateUserReq Decoder", request)
```
Comment (adilanchian) [inline cmd/createuser/createuser.go]: [nit]

`social.CreateUserReq Decoder` -> `CreateUserReq Decoder`

File: cmd/createuser/createuser.go [hunk 18]
```diff
@@ -34,15 +36,15 @@ func CreateUser(writer http.ResponseWriter, request *http.Request) {
...
 		return
 	}
 
 	// Get Document references for social platform
 	socialPlatDocRef := firestoreClient.Doc(createUserReq.SocialPlatformPath)
 	if socialPlatDocRef == nil {
 		http.Error(writer, jsonDecodeErr.Error(), http.StatusInternalServerError)
-		log.Printf("CreateUser [social.CreateUserReq Decoder]: %v", jsonDecodeErr)
+		logger.LogErr(jsonDecodeErr, "social.CreateUserReq Decoder", request)
```
Comment (adilanchian) [inline cmd/createuser/createuser.go]: [nit]

`social.CreateUserReq Decoder` -> `CreateUserReq Decoder`

File: cmd/createuser/createuser.go [hunk 19]
```diff
@@ -62,7 +64,7 @@ func CreateUser(writer http.ResponseWriter, request *http.Request) {
 	_, writeErr := firestoreClient.Collection("users").Doc(firestoreUser.ID).Set(context.Background(), firestoreUser)
 	if writeErr != nil {
 		http.Error(writer, writeErr.Error(), http.StatusInternalServerError)
-		log.Printf("CreateUser [fireStore Set]: %v", writeErr)
+		logger.LogErr(writeErr, "fireStore Set", nil)
```
Comment (adilanchian) [inline cmd/createuser/createuser.go]: [nit]

`fireStore` -> `FireStore`

File: cmd/doesuserdocexist/doesuserdocexist.go [hunk 20]
```diff
@@ -36,7 +38,7 @@ func DoesUserDocExist(writer http.ResponseWriter, request *http.Request) {
 	initWithEnvErr := initWithEnv()
 	if initWithEnvErr != nil {
 		http.Error(writer, initWithEnvErr.Error(), http.StatusInternalServerError)
-		log.Printf("DoesUserDocExist [initWithEnv]: %v", initWithEnvErr)
+		logger.LogErr(initWithEnvErr, "initWithEnv", nil)
```
Comment (adilanchian) [inline cmd/doesuserdocexist/doesuserdocexist.go]: [nit]

`initWithEnv` -> `InitWithEnv`

File: cmd/doesuserdocexist/doesuserdocexist.go [hunk 21]
```diff
@@ -46,7 +48,7 @@ func DoesUserDocExist(writer http.ResponseWriter, request *http.Request) {
 	reqDataErr := json.NewDecoder(request.Body).Decode(&reqData)
 	if reqDataErr != nil {
 		http.Error(writer, reqDataErr.Error(), http.StatusInternalServerError)
-		log.Printf("DoesUserDocExist [reqData Decoder]: %v", reqDataErr)
+		logger.LogErr(reqDataErr, "reqData Decoder", request)
```
Comment (adilanchian) [inline cmd/doesuserdocexist/doesuserdocexist.go]: [nit]

`reqData` -> `ReqData`

File: cmd/getspotifymedia/getspotifymedia.go [hunk 22]
```diff
@@ -104,7 +106,7 @@ func GetSpotifyMedia(writer http.ResponseWriter, request *http.Request) {
 	initWithEnvErr := initWithEnv()
 	if initWithEnvErr != nil {
 		http.Error(writer, initWithEnvErr.Error(), http.StatusInternalServerError)
-		log.Printf("GetSpotifyMedia [initWithEnv]: %v", initWithEnvErr)
+		logger.LogErr(initWithEnvErr, "initWithEnv", nil)
```
Comment (adilanchian) [inline cmd/getspotifymedia/getspotifymedia.go]: [nit]

`initiWithEnv` -> `InitWithEnv`

File: cmd/getspotifymedia/getspotifymedia.go [hunk 23]
```diff
@@ -113,15 +115,15 @@ func GetSpotifyMedia(writer http.ResponseWriter, request *http.Request) {
...
+		logger.LogErr(spotifyReqDecodeErr, "Request Decoder", request)
 		return
 	}
 
 	// Check to see if media is already part of collection, if so, just return that
 	mediaData, mediaDataErr := getMediaFromFirestore(spotifyMediaReq.MediaID)
 	if mediaDataErr != nil {
 		http.Error(writer, mediaDataErr.Error(), http.StatusInternalServerError)
-		log.Printf("GetSpotifyMedia [getMediaFromFirestore]: %v", mediaDataErr)
+		logger.LogErr(mediaDataErr, "getMediaFromFirestore", request)
```
Comment (adilanchian) [inline cmd/getspotifymedia/getspotifymedia.go]: [nit]

`getMediaFromFirestore` -> `GetMediaFromFirestore`

File: cmd/getspotifymedia/getspotifymedia.go [hunk 24]
```diff
@@ -137,7 +139,7 @@ func GetSpotifyMedia(writer http.ResponseWriter, request *http.Request) {
 	creds, credErr := getCreds()
 	if credErr != nil {
 		http.Error(writer, credErr.Error(), http.StatusInternalServerError)
-		log.Printf("GetSpotifyMedia [getCreds]: %v", credErr)
+		logger.LogErr(credErr, "getCreds", nil)
```
Comment (adilanchian) [inline cmd/getspotifymedia/getspotifymedia.go]: [nit]

`getCreds` -> `GetCreds`

File: cmd/getspotifymedia/getspotifymedia.go [hunk 25]
```diff
@@ -176,7 +178,7 @@ func GetSpotifyMedia(writer http.ResponseWriter, request *http.Request) {
 		firestoreMediaAlbumData, firestoreMediaDataErr := getSpotifyAlbum(spotifyMediaReq.MediaID, creds.AccessToken)
 		if firestoreMediaDataErr != nil {
 			http.Error(writer, firestoreMediaDataErr.Error(), http.StatusInternalServerError)
-			log.Printf("GetSpotifyMedia [GetSpotifyPlaylist Switch]: %v", firestoreMediaDataErr)
+			logger.LogErr(firestoreMediaDataErr, "GetSpotifyPlaylist Switch", request)
```
Comment (adilanchian) [inline cmd/getspotifymedia/getspotifymedia.go]: [nit]

`GetSpotifyPlaylist Switch ` -> `GetSpotifyAlbum Switch `

File: cmd/socialplatform/socialplatform.go [hunk 26]
```diff
@@ -26,7 +28,7 @@ func CreateSocialPlatform(writer http.ResponseWriter, request *http.Request) {
 	initWithEnvErr := initWithEnv()
 	if initWithEnvErr != nil {
 		http.Error(writer, initWithEnvErr.Error(), http.StatusInternalServerError)
-		log.Printf("CreateSocialPlatform [initWithEnv]: %v", initWithEnvErr)
+		logger.LogErr(initWithEnvErr, "initWithEnvErr", nil)
```
Comment (adilanchian) [inline cmd/socialplatform/socialplatform.go]: [nit]

`initWithEnvErr` -> `InitWithEnv`

File: cmd/socialtokenrefresh/socialtokenrefresh.go [hunk 27]
```diff
@@ -49,7 +51,7 @@ func SocialTokenRefresh(writer http.ResponseWriter, request *http.Request) {
 	initWithEnvErr := initWithEnv()
 	if initWithEnvErr != nil {
 		http.Error(writer, initWithEnvErr.Error(), http.StatusInternalServerError)
-		log.Printf("SocialTokenRefresh [initWithEnv]: %v", initWithEnvErr)
+		logger.LogErr(initWithEnvErr, "initWithEnv", nil)
```
Comment (adilanchian) [inline cmd/socialtokenrefresh/socialtokenrefresh.go]: [nit]

`initWithEnv` -> `InitWithEnv`

File: cmd/socialtokenrefresh/socialtokenrefresh.go [hunk 28]
```diff
@@ -60,15 +62,15 @@ func SocialTokenRefresh(writer http.ResponseWriter, request *http.Request) {
 	socialTokenErr := json.NewDecoder(request.Body).Decode(&socialTokenReq)
 	if socialTokenErr != nil {
 		http.Error(writer, socialTokenErr.Error(), http.StatusInternalServerError)
-		log.Printf("SocialTokenRefresh [socialTokenReq Decoder]: %v", socialTokenErr)
+		logger.LogErr(socialTokenErr, "socialTokenReq Decoder", request)
```
Comment (adilanchian) [inline cmd/socialtokenrefresh/socialtokenrefresh.go]: [nit]

`socialTokenReq` -> `SocialTokenReq`

File: cmd/socialtokenrefresh/socialtokenrefresh.go [hunk 29]
```diff
@@ -60,15 +62,15 @@ func SocialTokenRefresh(writer http.ResponseWriter, request *http.Request) {
...
 		return
 	}
 
 	// Go to Firestore and get the platforms for user
 	platsToRefresh, platformErr := getUserPlatformsToRefresh(socialTokenReq.UID)
 	if platformErr != nil {
 		http.Error(writer, platformErr.Error(), http.StatusInternalServerError)
-		log.Printf("SocialTokenRefresh [getUserPlatforms]: %v", platformErr)
+		logger.LogErr(platformErr, "getUserPlatforms", request)
```
Comment (adilanchian) [inline cmd/socialtokenrefresh/socialtokenrefresh.go]: [nit]

`getUserPlatforms` -> `GetUserPlatforms`

File: cmd/socialtokenrefresh/socialtokenrefresh.go [hunk 30]
```diff
@@ -105,6 +107,13 @@ func initWithEnv() error {
...
 	}
 
+	// Initialize Sawmill
+	sawmillLogger, err := sawmill.InitClient(currentProject, os.Getenv("GCLOUD_CONFIG"), os.Getenv("ENVIRONMENT"), "SocialTokenRefersh")
+	if err != nil {
+		log.Printf("SocialTokenRefresh [Init Sawmill]: %v", err)
+	}
+
+	logger = sawmillLogger
 	firestoreClient = client
```
Comment (adilanchian) [inline cmd/socialtokenrefresh/socialtokenrefresh.go]: You remove firestoreClient? Or am I reading this wrong?

File: cmd/spotifyauth/spotifyauth.go [hunk 31]
```diff
@@ -48,7 +50,7 @@ func AuthorizeWithSpotify(writer http.ResponseWriter, request *http.Request) {
 	initWithEnvErr := initWithEnv()
 	if initWithEnvErr != nil {
 		http.Error(writer, initWithEnvErr.Error(), http.StatusInternalServerError)
-		log.Printf("AuthorizeWithSpotify [initWithEnv]: %v", initWithEnvErr)
+		logger.LogErr(initWithEnvErr, "initWithEnvErr", nil)
```
Comment (adilanchian) [inline cmd/spotifyauth/spotifyauth.go]: [nit]

`initWithEnvErr` -> `initWithEnv`

File: cmd/spotifyauth/spotifyauth.go [hunk 32]
```diff
@@ -57,20 +59,20 @@ func AuthorizeWithSpotify(writer http.ResponseWriter, request *http.Request) {
 	authResponseErr := json.NewDecoder(request.Body).Decode(&spotifyAuthRequest)
 	if authResponseErr != nil {
 		http.Error(writer, authResponseErr.Error(), http.StatusInternalServerError)
-		log.Printf("AuthorizeWithSpotify [spotifyAuthRequest Decoder]: %v", authResponseErr)
+		logger.LogErr(authResponseErr, "spotifyAuthRequest Decoder", request)
```
Comment (adilanchian) [inline cmd/spotifyauth/spotifyauth.go]: [nit]

`spotifyAuthRequest` -> `SpotifyAuthRequest`

File: cmd/spotifyauth/spotifyauth.go [hunk 33]
```diff
@@ -57,20 +59,20 @@ func AuthorizeWithSpotify(writer http.ResponseWriter, request *http.Request) {
...
-		log.Println("AuthorizeWithSpotify: ApiToken was empty.")
+		logger.LogErr(fmt.Errorf("ApiToken was empty"), "spotifyAuthRequest Decoder", request)
 		return
 	}
 
 	spotifyMeReq, spotifyMeReqErr := http.NewRequest("GET", spotifyMeURI, nil)
 	if spotifyMeReqErr != nil {
 		http.Error(writer, spotifyMeReqErr.Error(), http.StatusInternalServerError)
-		log.Printf("AuthorizeWithSpotify [http.NewRequest]: %v", spotifyMeReqErr)
+		logger.LogErr(spotifyMeReqErr, "http.NewRequest", spotifyMeReq)
```
Comment (adilanchian) [inline cmd/spotifyauth/spotifyauth.go]: [nit]

`http.NewRequest` -> `Request`

File: cmd/spotifyauth/spotifyauth.go [hunk 34]
```diff
@@ -79,7 +81,7 @@ func AuthorizeWithSpotify(writer http.ResponseWriter, request *http.Request) {
 	resp, httpErr := httpClient.Do(spotifyMeReq)
 	if httpErr != nil {
 		http.Error(writer, httpErr.Error(), http.StatusBadRequest)
-		log.Printf("AuthorizeWithSpotify [client.Do]: %v", httpErr)
+		logger.LogErr(httpErr, "client.Do", spotifyMeReq)
```
Comment (adilanchian) [inline cmd/spotifyauth/spotifyauth.go]: [nit]

`client.Do` -> `GET Request`

File: cmd/spotifyauth/spotifyauth.go [hunk 35]
```diff
@@ -107,15 +109,15 @@ func AuthorizeWithSpotify(writer http.ResponseWriter, request *http.Request) {
...
+		logger.LogErr(respDecodeErr, "Spotify Request Decoder", spotifyMeReq)
 		return
 	}
 
 	// Check DB for user, if there return user object
 	authorizeWithSpotifyResp, userErr := getUser(spotifyMeResponse.ID)
 	if userErr != nil {
 		http.Error(writer, userErr.Error(), http.StatusBadRequest)
-		log.Printf("AuthorizeWithSpotify: %v", userErr)
+		logger.LogErr(userErr, "getUser", nil)
```
Comment (adilanchian) [inline cmd/spotifyauth/spotifyauth.go]: [nit]

`getUser` ->`GetUser`

File: cmd/spotifyauth/spotifyauth.go [hunk 36]
```diff
@@ -125,7 +127,7 @@ func AuthorizeWithSpotify(writer http.ResponseWriter, request *http.Request) {
 		socialPlatDocRef, socialPlatData, socialPlatErr := createSocialPlatform(spotifyMeResponse, spotifyAuthRequest)
 		if socialPlatErr != nil {
 			http.Error(writer, socialPlatErr.Error(), http.StatusBadRequest)
-			log.Printf("AuthorizeWithSpotify: [createSocialPlatform] %v", socialPlatErr)
+			logger.LogErr(socialPlatErr, "createSocialPlatform", nil)
```
Comment (adilanchian) [inline cmd/spotifyauth/spotifyauth.go]: [nit]

`createSocialPlatform` -> `CreateSocialPlatform`

File: cmd/spotifyauth/spotifyauth.go [hunk 37]
```diff
@@ -141,7 +143,7 @@ func AuthorizeWithSpotify(writer http.ResponseWriter, request *http.Request) {
 		var customToken, customTokenErr = getCustomToken(firestoreUser.ID)
 		if customTokenErr != nil {
 			http.Error(writer, customTokenErr.Error(), http.StatusBadRequest)
-			log.Printf("AuthorizeWithSpotify [customToken]: %v", customTokenErr)
+			logger.LogErr(customTokenErr, "customToken", nil)
```
Comment (adilanchian) [inline cmd/spotifyauth/spotifyauth.go]: [nit]

`customToken` -> `CustomToken`

File: cmd/spotifyauth/spotifyauth.go [hunk 38]
```diff
@@ -168,7 +170,7 @@ func AuthorizeWithSpotify(writer http.ResponseWriter, request *http.Request) {
 		var customToken, customTokenErr = getCustomToken(authorizeWithSpotifyResp.ID)
 		if customTokenErr != nil {
 			http.Error(writer, customTokenErr.Error(), http.StatusBadRequest)
-			log.Printf("AuthorizeWithSpotify [customToken]: %v", customTokenErr)
+			logger.LogErr(customTokenErr, "customToken", nil)
```
Comment (adilanchian) [inline cmd/spotifyauth/spotifyauth.go]: [nit]

`customToken` -> `CustomToken`

File: cmd/spotifyauth/spotifyauth.go [hunk 39]
```diff
@@ -201,9 +203,16 @@ func initWithEnv() error {
...
+	sawmillLogger, err := sawmill.InitClient(currentProject, os.Getenv("GCLOUD_CONFIG"), os.Getenv("ENVIRONMENT"), "AuthorizeWithSpotify")
+	if err != nil {
+		log.Printf("AuthorizeWithSpotify [Init Sawmill]: %v", err)
+	}
+
 	log.Println(currentProject)
 	log.Println(hostname)
 
+	logger = sawmillLogger
 	firestoreClient = client
```
Comment (adilanchian) [inline cmd/spotifyauth/spotifyauth.go]: Add `logger = sawmillLogger` after `firestoreClient = client`

File: cmd/usernameavailable/usernameavailable.go [hunk 40]
```diff
@@ -34,7 +36,7 @@ func UsernameAvailable(writer http.ResponseWriter, request *http.Request) {
 	initWithEnvErr := initWithEnv()
 	if initWithEnvErr != nil {
 		http.Error(writer, initWithEnvErr.Error(), http.StatusInternalServerError)
-		log.Printf("DoesUserDocExist [initWithEnv]: %v", initWithEnvErr)
+		logger.LogErr(initWithEnvErr, "initWithEnv", nil)
```
Comment (adilanchian) [inline cmd/usernameavailable/usernameavailable.go]: [nit]

`initWithEnv` -> `InitWithEnv`

File: cmd/usernameavailable/usernameavailable.go [hunk 41]
```diff
@@ -44,7 +46,7 @@ func UsernameAvailable(writer http.ResponseWriter, request *http.Request) {
 	reqDataErr := json.NewDecoder(request.Body).Decode(&reqData)
 	if reqDataErr != nil {
 		http.Error(writer, reqDataErr.Error(), http.StatusInternalServerError)
-		log.Printf("UsernameAvailable [reqData Decoder]: %v", reqDataErr)
+		logger.LogErr(reqDataErr, "reqData Decoder", request)
```
Comment (adilanchian) [inline cmd/usernameavailable/usernameavailable.go]: [nit]

`reqData` -> `ReqData`

File: internal/mock-config.yaml [hunk 42]
```diff
@@ -11,3 +11,4 @@ FIREBASE_PROJECTID_DEV: <AddRegisteredFirebaseProjectID>
 ALGOLIA_INDEX_NAME_DEV: <AddCreatedAlgoliaIndexName>
 APPLE_AUTH_TEMPLATE_PATH_DEV: cmd/appleauth/templates
 APPLE_PRIVATE_KEY_PATH_DEV: cmd/createappledevtoken/apple_private_key.p8
+GCLOUD_CONFIG: <ServiceAccountConfigJSONObjectPutItAllInOneLineAndWrapItWithSingleQuotes>
```
Comment (adilanchian) [inline internal/mock-config.yaml:14]: L.U.L.

Keep it for the memes kappa.

**Review by adilanchian — CHANGES_REQUESTED** [review]
First off - THANK YOU. This was an insane undertaking and I appreciate you taking the initiative to get this done. Grüvee thanks you <3

Everything looks fantastic. The comments I left were mainly so we don't forget to clean up certain spots. So yes, you will see 42 comments, but they are essentially all nits. 

There is one comment in regards to changing params around in Sawmill so please make sure you give that some thought.

I also wanted to add one comment about UpdateAlgolia. Is there a reason why you did not add Sawmill to that?

Thanks again!

File: cmd/appleauth/appleauth.go (hunk 2 above)
Comment (LeviHarrison) [reply cmd/appleauth/appleauth.go]: Sure! If that looks order looks better to you I'm glad to implement it.

File: cmd/socialtokenrefresh/socialtokenrefresh.go (hunk 30 above)
Comment (LeviHarrison) [reply cmd/socialtokenrefresh/socialtokenrefresh.go]: I think I just populated `sawmillClient` before `firestoreClient`. I'll fix that order.

**Comment by LeviHarrison** [comment]
Fixed up the capitalization issues, changed the order of the parameters for `LogErr`, and caught one very pesky typo. Regarding UpdateAlgolia, originally there were two ways to initialize a Sawmill Client, one which I used in most of the functions, and the other which I used in a few functions that didn't have `initWithEnv`. The problem was that the functions using the other way of initializing a Sawmill Client (including UpdateAlgolia), weren't working, so I assumed that it had to do with the functions. Later I found that way of initializing to be flawed, but I didn't put two and two together, so I still believed that UpdateAlgolia was just an anomaly.  I'll get around to adding Sawmill to it at some point, I just don't have my config stuff set up on this laptop right now so I can't test if it will load all the ENVs.

**Pull request merged**